File: backend/chess/pieces/piece.js
class Piece {
  constructor(x, y, color) {
    this.color = color;
    this.x = x;
    this.y = y;
    this.hasBeenMoved = false;
  }

  static fromSerialized(serialized) {
    if (serialized === " ") {
      return null;
    }
    color = (serialized.toUpperCase() === serialized) ? Colors.WHITE : Colors.BLACK;

    serialized = serialized.toLowerCase();
    switch (serialized) {
//      case "p":
//        return new Pawn(color);
      case "r":
        return new Rook(color);
      case "n":
        return new Knight(color);
      case "b":
        return new Bishop(color);
      case "q":
        return new Queen(color);
      case "k":
        return new King(color);
      default:
        throw new Error("Invalid serialized piece");
    }
  }
}

export default Piece;

File: rust-cli/chesslib/src/piece.rs
use crate::Color;

#[derive(Clone, Copy)]
pub enum Piece {
    Pawn,
    Knight,
    Bishop,
    Rook,
    Queen,
    King,
}

#[derive(Clone, Copy)]
pub struct ColoredPiece {
    color: Color,
    piece: Piece,
}
